modules/CbChargeSOMDriver/chargesom/CbChargeSOM.cpp

```diff
@@ -3,251 +3,681 @@
+}
+
+std::ostream& operator<<(std::ostream& os, enum estop_state state) {
+    return os << cb_proto_estop_state_to_str(state);
+}
+
+CbChargeSOM::CbChargeSOM() {
+    // clear the context structs before usage
+    memset(&this->uart, 0, sizeof(this->uart));
     memset(&this->ctx, 0, sizeof(this->ctx));
 
     // set to invalid fd so that destructor knows whether it must close something
-    this->ctx.uart.fd = -1;
+    this->uart.fd = -1;
+
+    // we need a thread to call the notification signals asynchronously to the
+    // frame receiving to avoid stalling and to not miss a single change
+    this->notify_thread = std::thread([&]() {
+        enum cp_state previous_cp_state = CP_STATE_MAX;
+        enum pp_state previous_pp_state = PP_STATE_MAX;
+        unsigned int previous_cp_errors = 0;
+        bool previous_contactor_error[CB_PROTO_MAX_CONTACTORS] = {};
+        bool previous_estop_tripped[CB_PROTO_MAX_ESTOPS] = {};
+
+        EVLOG_debug << "Notify Thread started";
+
+        while (!this->termination_requested) {
+            // temporary helper variables for our access functions
+            struct safety_controller tmpctx;
+            enum cp_state current_cp_state;
+            enum pp_state current_pp_state;
+            unsigned int current_cp_errors;
+            bool current_contactor_error[CB_PROTO_MAX_CONTACTORS];
+            bool current_estop_tripped[CB_PROTO_MAX_ESTOPS];
+            unsigned int i;
+
+            // wait for changes
+            std::unique_lock<std::mutex> lock(this->notify_mutex);
+            this->notify_cv.wait(lock, [&]() {
+                return (!this->charge_state_changes.empty() && this->evse_enabled) || this->termination_requested;
+            });
+
+            if (this->termination_requested)
+                break;
+
+            // remove from queue
+            tmpctx.charge_state = this->charge_state_changes.front();
+            this->charge_state_changes.pop();
+
+            //
+            // check for errors before doing the normal PP / CP state processing
+            // this give upper layer the chance to report the error and/or set flags
+            // before the actual low-level change is processed
+            //
+
+            // check for CP related errors
+            current_cp_errors = cb_proto_get_cp_errors(&tmpctx);
+            if (current_cp_errors != previous_cp_errors) {
+                EVLOG_debug << "on_cp_error(0x" << std::hex << std::setw(2) << std::setfill('0') << previous_cp_errors
+                            << " → "
+                            << "0x" << std::hex << std::setw(2) << std::setfill('0') << current_cp_errors << ")";
+                this->on_cp_error();
+                previous_cp_errors = current_cp_errors;
+            }
+
+            // forward contactor errors
+            for (i = 0; i < CB_PROTO_MAX_CONTACTORS; ++i) {
+                std::string name = "Contactor " + std::to_string(i + 1);
```

Comment:
It's a bit expensive to construct these strings CB_PROTO_MAX_CONTACTORS times on every notification, even if no contactor errors occurred.

Perhaps move it to the `if (current_contactor_error[i] != previous_contactor_error[i])` block scope, where it's actually used.